visivo/query/runner.py

```diff
@@ -60,73 +64,71 @@ def run(self):
             )
-        elif len(self.errors) > 0 and not self.soft_failure:
+        elif len(self.failed_job_results) > 0 and not self.soft_failure:
             Logger.instance().error(
-                f"\nRun failed in {round(time()-start_time, 2)}s with {len(self.errors)} query error(s)"
+                f"\nRun failed in {round(time()-start_time, 2)}s with {len(self.failed_job_results)} query error(s)"
             )
             exit(1)
         else:
             Logger.instance().info(f"\nRun finished in {round(time()-start_time, 2)}s")
 
-    def update_job_queue(self, job_tracker: JobTracker) -> bool:
-        all_dependencies_completed = True
-        for job in self.jobs:
-            job_item_children = all_descendants(dag=self.dag, from_node=job.item)
-            dependencies = list(
-                filter(
-                    lambda j: job != j and job_item_children.has_node(j.item),
-                    self.jobs,
-                )
-            )
-            incomplete_dependencies = list(
-                filter(
-                    lambda d: not job_tracker.is_job_name_done(d.name),
-                    dependencies,
-                )
-            )
-            dependencies_completed = len(incomplete_dependencies) == 0
-            if not dependencies_completed:
-                all_dependencies_completed = False
+    def job_callback(self, future: Future):
+        job_result: JobResult = future.result(timeout=0.01)
+        if job_result.success:
+            Logger.instance().success(str(job_result.message))
+            self.successful_job_results.append(job_result)
+        else:
+            Logger.instance().error(str(job_result.message))
+            self.failed_job_results.append(job_result)
+
+    def update_job_queue(self, job_tracker: JobTracker):
+        terminal_nodes = [
+            n for n in self.job_dag.nodes() if self.job_dag.out_degree(n) == 0
+        ]
 
-            if dependencies_completed and not job_tracker.is_job_name_enqueued(
-                job.name
+        failed_items = [result.item for result in self.failed_job_results]
+        successful_items = [result.item for result in self.successful_job_results]
+        for terminal_node in terminal_nodes:
+            descendants = all_descendants(dag=self.project_dag, from_node=terminal_node)
+            if terminal_node in successful_items or terminal_node in failed_items:
+                self.job_dag.remove_node(terminal_node)
+                continue
+            elif any(descendant in failed_items for descendant in descendants):
+                Logger.instance().info(
+                    f"Skipping job for {terminal_node} because it has a failed dependency"
+                )
+                self.job_dag.remove_node(terminal_node)
+                continue
+            elif self.name_filter and not family_tree_contains_named_node(
```

Comment:
So this is where we'd implement the dbt selector style syntax?

Reply:
Correct!